### PR TITLE
feat(plonky2x): hint witness generation

### PR DESCRIPTION
**Pull request by tamirhemo**
Creates a simple generator from a fn. 

Users can register hint fn on a witness registry using the `register_hint()` method.

### CODE REVIEW

File: plonky2x/src/frontend/eth/beacon/builder.rs [hunk 1]
```diff
@@ -217,7 +217,7 @@ impl<F: RichField + Extendable<D>, const D: usize> CircuitBuilder<F, D> {
             withdrawals,
             idx,
         );
-        self.add_simple_generator(&generator);
+        self.add_simple_generator(generator.clone());
```
Comment (puma314) [inline plonky2x/src/frontend/eth/beacon/builder.rs:220]: Why did you change all of this?

File: plonky2x/src/backend/circuit/serialization.rs [hunk 2]
```diff
@@ -200,6 +201,24 @@ impl<F: RichField + Extendable<D>, const D: usize> WitnessGeneratorRegistry<F, D
     pub fn register_simple<SG: SimpleGenerator<F, D>>(&mut self, id: String) {
         self.register::<SimpleGeneratorAdapter<F, SG, D>>(id)
     }
+
+    pub fn register_hint(&mut self, hint_fn: fn(&mut ValueStream<F, D>, &mut ValueStream<F, D>)) {
```
Comment (jtguibas) [inline plonky2x/src/backend/circuit/serialization.rs:205]: Should we have ReadableValueStream and WriteableValueStream? Right now it looks like you could read from the second stream.

File: plonky2x/src/frontend/vars/buffer.rs [hunk 3]
```diff
@@ -1,27 +1,173 @@
...
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Stream<T> {
+    data: Vec<T>,
+    position: usize,
+}
+
+impl<T> Stream<T> {
```
Comment (puma314) [inline plonky2x/src/frontend/vars/buffer.rs:55]: If `Stream` has nothing to do with `Vars`, which it seems like it doesn't would prefer to maybe put it in `vars/utils` to make it clear that itself is not a var? 

Also I'm surprised there's no existing Rust type for this...

File: plonky2x/src/backend/circuit/serialization.rs (hunk 2 above)
Comment (puma314) [reply plonky2x/src/backend/circuit/serialization.rs:205]: I agree with this

File: plonky2x/src/frontend/generator/simple.rs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    fn run_once(&self, witness: &PartitionWitness<F>, out_buffer: &mut GeneratedValues<F>) {
+        self.generator.run(witness, out_buffer);
+    }
+
+    fn serialize(&self, dst: &mut Vec<u8>, _common_data: &CommonCircuitData<F, D>) -> IoResult<()> {
+        let bytes = bincode::serialize(self).map_err(|_| IoError)?;
+        dst.write_all(&bytes)
+    }
+
+    fn deserialize(src: &mut Buffer, _common_data: &CommonCircuitData<F, D>) -> IoResult<Self>
```
Comment (puma314) [inline plonky2x/src/frontend/generator/simple.rs:61]: It would be great to have a test here as an example of what adding a `Generator` is like?

File: plonky2x/src/frontend/generator/hint.rs [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::backend::circuit::serialization::{GateRegistry, WitnessGeneratorRegistry};
+    use crate::prelude::{ByteVariable, *};
+
+    fn plus_one<F: RichField + Extendable<D>, const D: usize>(
+        input_stream: &mut ValueStream<F, D>,
```
Comment (puma314) [inline plonky2x/src/frontend/generator/hint.rs:142]: Can you put all this stuff in the prelude so that our users have an easy time importing it?

File: plonky2x/src/frontend/generator/hint.rs [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+        circuit.verify(&proof, &input, &output);
+
+        // Read output.
+        let byte_plus_one = output.read::<ByteVariable>();
+        assert_eq!(byte_plus_one, 6u8);
+
+        // Test the serialization
+        let gate_serializer = GateRegistry::new();
+        let mut generator_serializer = WitnessGeneratorRegistry::new::<PoseidonGoldilocksConfig>();
+        generator_serializer.register_hint(plus_one);
```
Comment (puma314) [inline plonky2x/src/frontend/generator/hint.rs:182]: Do we have to `register_hint` on the `circuit` anywhere?

File: plonky2x/src/frontend/vars/buffer.rs [hunk 7]
```diff
@@ -1,27 +1,173 @@
...
+use serde::{Deserialize, Serialize};
+
 use super::{CircuitVariable, Variable};
-pub struct VariableBuffer<'a> {
-    variables: &'a [Variable],
-    position: usize,
+use crate::prelude::CircuitBuilder;
+
+#[derive(Debug, Clone)]
+pub struct ValueStream<F, const D: usize>(Stream<F>);
```
Comment (puma314) [inline plonky2x/src/frontend/vars/buffer.rs:12]: can we add comments on what each of these things are meant to be?

File: plonky2x/src/frontend/eth/beacon/builder.rs (hunk 1 above)
Comment (jtguibas) [reply plonky2x/src/frontend/eth/beacon/builder.rs:220]: You shouldn't use the generator after you've added it

File: plonky2x/src/frontend/vars/buffer.rs [hunk 8]
```diff
@@ -1,27 +1,173 @@
+use core::marker::PhantomData;
```
Comment (jtguibas) [inline plonky2x/src/frontend/vars/buffer.rs:1]: rename this file to stream?

File: plonky2x/src/frontend/vars/buffer.rs (hunk 3 above)
Comment (jtguibas) [reply plonky2x/src/frontend/vars/buffer.rs:55]: I feel like it shouldn't even be in `vars`, it should be in `utils` because it's used for ValueStream.

File: plonky2x/src/frontend/vars/buffer.rs [hunk 9]
```diff
@@ -1,27 +1,173 @@
...
+        variables
+    }
+    pub fn read<V: CircuitVariable>(&self, builder: &mut CircuitBuilder<F, D>) -> V {
+        let variables = self.read_exact(builder, V::nb_elements());
+        V::from_variables(&variables)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Stream<T> {
```
Comment (jtguibas) [inline plonky2x/src/frontend/vars/buffer.rs:50]: If we have ValueStream, can we also have VariableStream instead of Stream<Variable>?

File: plonky2x/src/frontend/eth/storage/vars/storage.rs [hunk 10]
```diff
@@ -198,7 +198,7 @@ impl CircuitVariable for EthLogVariable {
     }
 
     fn from_variables(variables: &[Variable]) -> Self {
-        let mut var_buffer = VariableBuffer::new(variables);
+        let mut var_buffer = Stream::<Variable>::new(variables.to_vec());
```
Comment (jtguibas) [inline plonky2x/src/frontend/eth/storage/vars/storage.rs:201]: Use VariableStream instead?

File: plonky2x/src/frontend/generator/hint.rs (hunk 6 above)
Comment (tamirhemo) [reply plonky2x/src/frontend/generator/hint.rs]: No, but users will have do return their own `GeneratorRegistry` when they implement CircuitFunction.

File: plonky2x/src/backend/circuit/serialization.rs (hunk 2 above)
Comment (tamirhemo) [reply plonky2x/src/backend/circuit/serialization.rs]: I named them ValueReader and ValueWriter

File: plonky2x/src/frontend/generator/simple.rs (hunk 4 above)
Comment (tamirhemo) [reply plonky2x/src/frontend/generator/simple.rs:62]: I can add an example without variables in it, but until we implement serde::Serialize and Deserialize on variable it's kind of blocked, so maybe this should be a separate PR?

**Review by jtguibas — APPROVED** [review]
